File: leetcode/pythonSolutions/daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        ans = [0] * len(temperatures)
        monostack = []

        for i in range(len(temperatures) - 1, -1, -1):
            while monostack and temperatures[monostack[-1]] <= temperatures[i]:
                monostack.pop()
            
            if monostack:
                ans[i] = monostack[-1] - i
            
            monostack.append(i)

        return ans

File: leetcode/C++Solutions/daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        vector<int> ans(temperatures.size(), 0);
        stack<int> monostack;

        for (int i = temperatures.size() - 1; i >= 0; --i) {
            while (!monostack.empty() && temperatures[monostack.top()] <= temperatures[i]) {
                monostack.pop();
            }

            if (!monostack.empty()) {
                ans[i] = monostack.top() - i;
            }

            monostack.push(i);
        }

        return ans;
    }
};
